File: macos/MISILNative/Sources/MISILNative/AgerbotManagedRuntimeActivator.swift
import Foundation

@MainActor
final class AgerbotManagedRuntimeActivator: AgerbotRuntimeActivating {
    private let processManager: AgerbotProcessManager
    private let settingsStore: AgerbotSettingsStore
    private let installationManager: AgerbotInstallationManager

    init(
        processManager: AgerbotProcessManager,
        settingsStore: AgerbotSettingsStore,
        installationManager: AgerbotInstallationManager
    ) {
        self.processManager = processManager
        self.settingsStore = settingsStore
        self.installationManager = installationManager
    }

    func activate(_ record: AgerbotCurrentModelRecord) async -> Bool {
        await processManager.stop()
        settingsStore.settings.checkpointPath = record.checkpointPath
        settingsStore.settings.activeModelVersion = record.activeVersion
        settingsStore.settings.previousModelVersion = record.previousVersion
        installationManager.refresh(using: settingsStore.settings)
        await processManager.start(
            settings: settingsStore.settings,
            installationStatus: installationManager.status
        )
        guard case let .ready(health) = processManager.state else { return false }
        return health.model.version == record.activeVersion
    }
}

File: windows/MISILNative.Core/Agerbot/AgerbotManagedRuntimeActivator.cs
namespace MISILNative.Core.Agerbot;

public sealed class AgerbotManagedRuntimeActivator : IAgerbotModelRuntimeActivator
{
    private readonly IAgerbotProcessManager _processManager;
    private readonly AgerbotSettingsStore _settingsStore;

    public AgerbotManagedRuntimeActivator(
        IAgerbotProcessManager processManager,
        AgerbotSettingsStore settingsStore)
    {
        _processManager = processManager;
        _settingsStore = settingsStore;
    }

    public async Task<bool> ActivateAsync(AgerbotCurrentModelRecord record, CancellationToken cancellationToken = default)
    {
        await _processManager.StopAsync(cancellationToken);
        _settingsStore.Update(settings =>
        {
            settings.CheckpointPath = record.CheckpointPath;
            settings.ActiveModelVersion = record.ActiveVersion;
            settings.PreviousModelVersion = record.PreviousVersion;
        });
        await _processManager.StartAsync(_settingsStore.Settings, cancellationToken);
        return _processManager.State.IsReady
            && _processManager.State.Health?.Model.Version == record.ActiveVersion;
    }

    public async Task DeactivateAsync(CancellationToken cancellationToken = default)
    {
        await _processManager.StopAsync(cancellationToken);
        _settingsStore.Update(settings =>
        {
            settings.CheckpointPath = string.Empty;
            settings.ActiveModelVersion = null;
            settings.PreviousModelVersion = null;
        });
    }
}
